## Scripts Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum### tool.bk.bam\_to\_ctss.plThis tool converts a bulk CAGE bam file to a ctss bed file, identifies read 5'end (capped TSS, i.e. ctss), extracts the unencoded G information, pileup ctss, and deduplicate the UMI``` tool.bk.bam_to_ctss.pl [options] --bamPath --genome --outputPrefix --outDir  Required:   --bamPath       [required] [string]  bam file (of CAGE reads), can be read 1 only or pair-end   --genome        [required] [string]  name of genome reference, e.g. hg19.gencode_v32lift37   --outputPrefix  [required] [string]  prefix for the output files   --outDir        [required] [string]  directory for the output files Optional:   --include_flag  [optional] [string]  samflag to be included, comma delimited                                         e.g. '64' to include read1, (default=null)   --exclude_flag  [optional] [string]  samflag to be excluded, comma delimited,                                         e.g. '128,256,4' to exclude read2, secondary alignment                                         and unaligned reads (default=128,256,4)   --min_MAPQ      [optional] [integer] minimum MAPQ to include (default=0)   --max_thread    [optional] [integer] maximum number of parallel threads, capped at 10 to                                         avoid memory overflow (default=5)   --overwrite     [optional] [yes/no]  erase outDir/outputPrefix before running (default=no)```### tool.bk.count.plThis tool counts the CAGE reads within a set of user-defined regions, e.g. tCRE, and returns the reads per regions``` tool.bk.count.pl [options] --countRegion_bed_path --ctss_bed_path --outputPrefix --outDir  Required:   --countRegion_bed_path   [required] [string] bed file contains the regions for counting CTSS, e.g. tCRE ranges,                                                 *.CRE.coord.bed.gz from tool.cm.annotate.pl   --ctss_bed_path          [required] [string] ctss file for counting,                                                *CB.ctss.bed.gz from tool.sc.bam_to_ctss.pl,                                                 4th column cellbarcode and 5th column is number UMI   --outputPrefix           [required] [string] prefix for the output files   --outDir                 [required] [string] directory for the output files Optional:   --overwrite              [optional] [yes/no]  erase outDir/outputPrefix before running (default=no)```### tool.bk.pool.plThis tool pools multiple bulk CAGE ctss bed file``` tool.bk.pool.pl --lib_list_path --genome --outputPrefix --outDir Required:   --lib_list_path [required] [string]  a list of libraries, in formation of                                         <lib_ID><\t><long_ctss_bed><\t><collapse_ctss_bed>                                        lib_ID = Unique ID of the cellbarcode                                        long_ctss_bed = *long.ctss.bed.gz from tool.bk.bam_to_ctss.pl,                                         collapse_ctss_bed = *collapse.ctss.bed.gz from tool.bk.bam_to_ctss.pl,    --genome        [required] [string]  name of genome reference, e.g. hg19.gencode_v32lift37   --outputPrefix  [required] [string]  prefix for the output files   --outDir        [required] [string]  directory for the output files Optional:   --max_thread    [optional] [integer] maximum number of parallel threads, capped at 10 to                                         avoid memory overflow (default=5)   --overwrite     [optional] [yes/no]  erase outDir/outputPrefix before running (default=no)```### tool.bk.subsample_ctss.plThis tool subsample a ctss bed file from bulk CAGE ctss``` tool.bk.subsample_ctss.pl --UMI_CB_ctss_bed_path --subsample_num --outputPrefix --outDir Required:   --long_ctss_bed_path  [required] [string]  ctss file for subsampling, one line read in "long" format,                                               *long.ctss.bed.gz from tool.bk.bam_to_ctss.pl,    --subsample_num       [required] [integer] number of UMI to be subsampled   --outputPrefix        [required] [string]  prefix for the output files   --outDir              [required] [string]  directory for the output files Optional:   --overwrite           [optional] [yes/no]  erase outDir/outputPrefix before running (default=no)```### tool.cm.annotate.plThis tool defines tCRE from TSS clusters and annotates them based their overlap with gene models.``` tool.cm.annotate.pl [options] --tssCluster_bed_path --tssCluster_info_path --genome --outputPrefix --outDir Required:   --tssCluster_bed_path      [required] [string]  bed file contains the ranges of filtered TSS clusters,                                                   *.tssCluster.*.filtered.bed.gz from tool.cm.filter.pl   --tssCluster_info_path     [required] [string]  tsv file contains the information of all TSS clusters,                                                   *.tssCluster.log.tsv from tool.cm.filter.pl   --genome                   [required] [string]  name of genome reference, e.g. hg19.gencode_v32lift37   --outputPrefix             [required] [string]  prefix for the output files   --outDir                   [required] [string]  directory for the output files Optional:   --up_end5Rng               [optional] [integer] TSS clusters will be classified as gene TSS, exonic, intron                                                    and intergenic. $up_end5Rng determines the range upstream of                                                    annotated gene TSS to be used for gene TSS assignment                                                    (default = 500)   --dn_end5Rng               [optional] [integer] TSS clusters will be classified as gene TSS, exonic, intron                                                    and intergenic. $dn_end5Rng determines the range downstream of                                                    annotated gene TSS to be used for gene TSS assignment                                                    (default = 500)   --exon_slop_rng            [optional] [integer] TSS clusters will be classified as gene TSS, exonic, intron                                                    and intergenic. $exon_slop_rng determines the range to be extended                                                   (i.e. slopped) from exon for assignment of exonic class.                                                    Used -1 to NOT to extend (default = -1)   --merge_dist               [optional] [integer] TSS clusters outside annotated gene promoters are grouped                                                   as "dummy genes" (for operational uniformity) by merging closely                                                    located TSS clusters.  $merge_dist determines the maximum distances                                                    between TSS clusters to be merged (default = 500)   --addon_length             [optional] [integer] see $merge_dist. add-on "dummy transcrips" will assigned to TSS cluster of                                                    "dummy genes" (for operational uniformity).$addon_length determines                                                    the length of these add-on "dummy transcrips" (default = 500).   --proximity_slop_rng       [optional] [integer] TSS clusters will be assigned to annotated gene TSS are "proximal"                                                   TSS clusters. $proximity_slop_rng determines the range to be extended                                                   (i.e. slopped) from gene TSS for assignment of proximal TSS clusters.                                                    (default = 500)   --merge_strandness         [optional] [string]  see $merge_dist. $merge_strandness decides the merge to be                                                    strand-aware ("stranded") or strand-agnostic "strandless".                                                   (default = strandless)   --proximal_strandness      [optional] [string]  closely located proximal TSS clusters are merged                                                     tCREs. $proximal_strandness decides the merge to be                                                    strand-aware ("stranded") or strand-agnostic "strandless".                                                   (default = stranded)   --CRE_extend_size          [optional] [integer] tCREs were defined by merging the extended ranges of TSS clusters.                                                   $CRE_extend_size determine the size of this range (both sides of                                                    summit) (default = 500)   --CRE_extend_upstrm_ratio  [optional] [float]   see $CRE_extend_size. $CRE_extend_upstrm_ratio determines the ratio                                                    (X:1) of flanking sizes on the upstream and downstream of summit.                                                    e.g. $CRE_extend_upstrm_ratio=4, upstream and downstream size will be                                                    taken as 4:1 ratio. $CRE_extend_size=500 and $CRE_extend_upstrm_ratio=4,                                                   upstream and downstream will be 400 and 100 respectively                                                    (default = 4)   --overwrite                [optional] [yes/no]  erase outDir/outputPrefix before running (default=no)```### tool.cm.cluster.plThis tool generate TSS cluster from a ctss bed file, using an external tool paraclu with user-defined cutoffs``` tool.cm.cluster.pl [options] --cluster_ctss_bed_path --outputPrefix --outDir Required:   --cluster_ctss_bed_path        [required] [string]  ctss file used for clustering,                                                       "collapse" ctss file from tool.sc.bam_to_ctss.pl,                                                        4th column is number of cells and 5th column is number UMI   --outputPrefix                 [required] [string]  prefix for the output files   --outDir                       [required] [string]  directory for the output files Optional:   --count_ctss_bed_path_list     [optional] [string]  comma delimited list of ctss bed file,                                                        using for filtering of clusters based signal                                                        (default=$cluster_ctss_bed_path)    --count_scope_bed_path         [optional] [string]  a bed file specify the scope for counting in $count_ctss_bed_path_list,                                                        using for filtering of clusters based signal                                                       (default=$cluster_ctss_bed_path)    --min_pos_count                [optional] [integer] minimum counts per position, used for filtering the raw signal                                                        in $cluster_ctss_bed_path before clustering (default = 1)   --min_cluster_cpm              [optional] [float]   minimum counts per million (cpm) for a cluster (default = 1e-5)   --min_summit_count             [optional] [integer] minimum counts at the summit of a cluster (default = 3)   --min_cluster_count            [optional] [integer] minimum counts within a cluster (default = 5)   --min_num_sample_expr_cluster  [optional] [integer] minimum number of samples (or cells) detected at the                                                        summit of a cluster (default = 3)   --min_num_sample_expr_summit   [optional] [integer] minimum number of samples (or cells) detected within                                                        of a cluster (default = 5)   --merge_dist                   [optional] [integer] maximum distance for merging closely located clusters,                                                        -1 to turn off merging (default = -1)   --overwrite                    [optional] [yes/no]  erase outDir/outputPrefix before running (default=no)```###tool.cm.ctss\_to\_bigwig.plThis tool converts a ctss bed file into two bigwig file, one for each strand, for visualization purpose    ``` tool.cm.ctss_to_bigwig.pl [options] --ctss_bed_path --genome --outputPrefix --outDir Required:   --ctss_bed_path      [required] [string] "collapse" ctss file from tool.sc.bam_to_ctss.pl   --genome             [required] [string] name of genome reference, e.g. hg19.gencode_v32lift37   --outputPrefix       [required] [string] prefix for the output files   --outDir             [required] [string] directory for the output files Optional:   --overwrite          [optional] [yes/no] erase outDir/outputPrefix before running (default=no)```### tool.cm.filter.plThis tool filter for genuine TSS clusters based on a logistic regression model on TSS cluster properties.```   tool.cm.filter.pl [options] --ctss_bed_path --ung_ctss_bed_path --tssCluster_bed_path --genome --outputPrefix --outDir```### tool.cm.prep_genome.plThis tool prepares a reference genome assembly and its gene models for others tools in scafe.```   tool.cm.prep_genome.pl [options] --gtf_path --fasta_path --chrom_list_path --mask_bed_path --outputPrefix --outDir```### tool.cm.remove\_strand\_invader.plThis tool identify and remove strand invader artefact from a ctss bed file, by aligning the sequence immediate upstream of a ctss to TS oligo sequence   ```   tool.cm.remove_strand_invader.pl [options] --ctss_bed_path --genome --outputPrefix --outDir```### tool.sc.bam\_to\_ctss.plThis tool converts a bam file to a ctss bed file, identifies read 5'end (capped TSS, i.e. ctss), extracts the unencoded G information, pileup ctss, and deduplicate the UMI   ```   tool.sc.bam_to_ctss.pl [options] --bamPath --genome --outputPrefix --outDir```### tool.sc.count.pl      This tool counts the UMI within a set of user-defined regions, e.g. tCRE, and returns a UMI/cellbarcode matrix```   tool.sc.count.pl [options] --countRegion_bed_path --cellBarcode_list_path --ctss_bed_path --outputPrefix --outDir```### tool.sc.pool.plThis tool pool multiple ctss bed file and maintains the unique (suffixed) cellbarcode and UMI information```   tool.sc.pool.pl --lib_list_path --genome --outputPrefix --outDir```### tool.sc.subsample\_ctss.plThis tool subsample a ctss bed file and maintains the cellbarcode and UMI information```   tool.sc.subsample_ctss.pl --UMI_CB_ctss_bed_path --subsample_num --outputPrefix --outDir```### workflow.bk.pool.plThis workflow process a single sample, from a bulk CAGE bam file to read count per tCRE per sample```   workflow.sc.pool.pl [options] --lib_list_path --genome --run_tag --run_outDir```### workflow.bk.solo.plThis workflow process a single sample, from a bulk CAGE bam file to read count per tCRE ```   workflow.sc.solo.pl [options] --run_bam_path --genome --run_tag --run_outDir```### workflow.bk.subsample.plThis workflow subsamples a ctss file, defines tCRE and generate tCRE read count. Subsampling is useful to investigate the effect of sequencing depth to tCRE definition```   workflow.bk.subsample.pl [options] --long_ctss_bed_path --subsample_num --genome --run_tag --run_outDir```### workflow.sc.pool.plThis workflow process a single sample, from a cellranger bam file to tCRE UMI/cellbarcode count matrix```   workflow.sc.pool.pl [options] --lib_list_path --genome --run_tag --run_outDir```### workflow.sc.solo.plThis workflow process a single sample, from a cellranger bam file to tCRE UMI/cellbarcode count matrix```   workflow.sc.solo.pl [options] --run_bam_path --run_cellbarcode_path --genome --run_tag --run_outDir```### workflow.sc.subsample.plThis workflow subsamples a ctss file, defines tCRE and generate a tCRE UMI/cellbarcode count matrix. Subsampling is useful to investigate the effect of sequencing depth to tCRE definition```   workflow.sc.subsample.pl [options] --UMI_CB_ctss_bed_path --run_cellbarcode_path --subsample_num --genome --run_tag --run_outDir```