## Scripts Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum###tool.bk.bam\_to\_ctss.plThis tool converts a bulk CAGE bam file to a ctss bed file, identifies read 5'end (capped TSS, i.e. ctss), extracts the unencoded G information, pileup ctss, and deduplicate the UMI```   tool.bk.bam_to_ctss.pl [options] --bamPath --genome --outputPrefix --outDir```###tool.bk.count.pl   This tool counts the CAGE reads within a set of user-defined regions, e.g. tCRE, and returns the reads per regions```   tool.bk.count.pl [options] --countRegion_bed_path --ctss_bed_path --outputPrefix --outDir```###tool.bk.pool.pl   This tool pools multiple bulk CAGE ctss bed file```   tool.bk.pool.pl --lib_list_path --genome --outputPrefix --outDir```###tool.bk.subsample_ctss.pl   This tool subsample a ctss bed file from bulk CAGE ctss```   tool.sc.subsample_ctss.pl --UMI_CB_ctss_bed_path --subsample_num --outputPrefix --outDir```###tool.cm.annotate.pl   This tool defines tCRE from TSS clusters and annotates them based their overlap with gene models.```   tool.cm.annotate.pl [options] --tssCluster_bed_path --tssCluster_info_path --genome --outputPrefix --outDir```###tool.cm.cluster.pl   This tool generate TSS cluster from a ctss bed file, using an external tool paraclu with user-defined cutoffs```   tool.cm.cluster.pl [options] --cluster_ctss_bed_path --outputPrefix --outDir```###tool.cm.ctss\_to\_bigwig.pl   This tool converts a ctss bed file into two bigwig file, one for each strand, for visualization purpose    ```   tool.cm.ctss_to_bigwig.pl [options] --ctss_bed_path --genome --outputPrefix --outDir```###tool.cm.filter.pl   This tool filter for genuine TSS clusters based on a logistic regression model on TSS cluster properties.```   tool.cm.filter.pl [options] --ctss_bed_path --ung_ctss_bed_path --tssCluster_bed_path --genome --outputPrefix --outDir```###tool.cm.prep_genome.pl   This tool prepares a reference genome assembly and its gene models for others tools in scafe.```   tool.cm.prep_genome.pl [options] --gtf_path --fasta_path --chrom_list_path --mask_bed_path --outputPrefix --outDir```###tool.cm.remove\_strand\_invader.pl   This tool identify and remove strand invader artefact from a ctss bed file,    by aligning the sequence immediate upstream of a ctss to TS oligo sequence   ```   tool.cm.remove_strand_invader.pl [options] --ctss_bed_path --genome --outputPrefix --outDir```###tool.sc.bam\_to\_ctss.pl   This tool converts a bam file to a ctss bed file, identifies read 5'end (capped TSS, i.e. ctss),   extracts the unencoded G information, pileup ctss, and deduplicate the UMI   ```   tool.sc.bam_to_ctss.pl [options] --bamPath --genome --outputPrefix --outDir```###tool.sc.count.pl   This tool counts the UMI within a set of user-defined regions, e.g. tCRE, and returns a UMI/cellbarcode matrix```   tool.sc.count.pl [options] --countRegion_bed_path --cellBarcode_list_path --ctss_bed_path --outputPrefix --outDir```###tool.sc.pool.pl   This tool pool multiple ctss bed file and maintains the unique (suffixed) cellbarcode and UMI information```   tool.sc.pool.pl --lib_list_path --genome --outputPrefix --outDir```###tool.sc.subsample\_ctss.pl   This tool subsample a ctss bed file and maintains the cellbarcode and UMI information```   tool.sc.subsample_ctss.pl --UMI_CB_ctss_bed_path --subsample_num --outputPrefix --outDir```###workflow.bk.pool.pl   This workflow process a single sample, from a bulk CAGE bam file to read count per tCRE per sample```   workflow.sc.pool.pl [options] --lib_list_path --genome --run_tag --run_outDir```###workflow.bk.solo.pl   This workflow process a single sample, from a bulk CAGE bam file to read count per tCRE ```   workflow.sc.solo.pl [options] --run_bam_path --genome --run_tag --run_outDir```###workflow.bk.subsample.pl   This workflow subsamples a ctss file, defines tCRE and generate tCRE read count   Subsampling is useful to investigate the effect of sequencing depth to tCRE definition```   workflow.bk.subsample.pl [options] --long_ctss_bed_path --subsample_num --genome --run_tag --run_outDir```###workflow.sc.pool.pl   This workflow process a single sample, from a cellranger bam file to tCRE UMI/cellbarcode count matrix```   workflow.sc.pool.pl [options] --lib_list_path --genome --run_tag --run_outDir```###workflow.sc.solo.pl   This workflow process a single sample, from a cellranger bam file to tCRE UMI/cellbarcode count matrix```   workflow.sc.solo.pl [options] --run_bam_path --run_cellbarcode_path --genome --run_tag --run_outDir```###workflow.sc.subsample.pl   This workflow subsamples a ctss file, defines tCRE and generate a tCRE UMI/cellbarcode count matrix   Subsampling is useful to investigate the effect of sequencing depth to tCRE definition```   workflow.sc.subsample.pl [options] --UMI_CB_ctss_bed_path --run_cellbarcode_path --subsample_num --genome --run_tag --run_outDir```